api/v1beta1/azuremachine_default.go

```diff
@@ -143,15 +158,43 @@ func (s *AzureMachineSpec) SetNetworkInterfacesDefaults() {
 	}
 }
 
+func (m *AzureMachine) getSubscriptionID(cli client.Client) (string, error) {
+	ctx := context.Background()
+
+	// Fetch the Cluster.
+	clusterName, ok := m.Labels[clusterv1.ClusterLabelName]
+	if !ok {
+		return "", nil
+	}
+
+	ownerCluster := &AzureCluster{}
+	key := client.ObjectKey{
+		Namespace: m.Namespace,
+		Name:      clusterName,
+	}
+
+	if err := cli.Get(ctx, key, ownerCluster); err != nil {
+		return "", err
+	}
+
+	return ownerCluster.Spec.SubscriptionID, nil
+}
+
 // SetDefaults sets to the defaults for the AzureMachineSpec.
-func (s *AzureMachineSpec) SetDefaults() {
-	if err := s.SetDefaultSSHPublicKey(); err != nil {
+func (m *AzureMachine) SetDefaults(client client.Client) {
+	if err := m.Spec.SetDefaultSSHPublicKey(); err != nil {
 		ctrl.Log.WithName("SetDefault").Error(err, "SetDefaultSshPublicKey failed")
 	}
-	s.SetDefaultCachingType()
-	s.SetDataDisksDefaults()
-	s.SetIdentityDefaults()
-	s.SetSpotEvictionPolicyDefaults()
-	s.SetDiagnosticsDefaults()
-	s.SetNetworkInterfacesDefaults()
+
+	subscriptionID, err := m.getSubscriptionID(client)
+	if err != nil {
+		ctrl.Log.WithName("SetDefault").Error(err, "getSubscriptionID failed")
```

Comment:
I observed above this line that the default SSH Public key uses the same error handling pattern. The default webhook function that calls `SetDefaults` does not seem to return an error. Should I handle this one differently?

Reply:
I think returning an error is preferable whenever that's possible

Reply:
The parent `Default` function actually doesn't return an error. I refactored the `SetDefault` function to return an error. Where should I log the returned error after it is passed to the webhook `Default` function?

Reply:
@CecileRobertMichon After making this change, it looks like the default function is returning before some of the other necessary defaults are set, particularly during unit tests that reference the default function. I did not observe any other Default webhook functions returning errors. Is there a standard for handling errors within webhooks? 

For example, the [`AzureMachineTemplate` webhook](https://github.com/willie-yao/cluster-api-provider-azure/blob/28eaa3755e32b6aa226a7e1a9d5baa99d0c399b3/api/v1beta1/azuremachinetemplate_webhook.go#L148) uses the `ctrl.log` pattern and doesn't return immediately if an error occurs.

Reply:
@willie-yao I think I missed this part before

> The default webhook function that calls SetDefaults does not seem to return an error

That's a tricky one because we don't want to prevent the other defaults from being set, in order to handle a potential failure we could best effort default (log error if error occurs but proceed with the rest of the func) then defer to the validation webhook to enforce that `getSubscriptionID` is set and fail if it isn't.

Reply:
I think that's a better way to go about it. I noticed that this prevented other defaults from being set too, so I'll go ahead and defer to the validation webhook.

Reply:
I made this [change](https://github.com/willie-yao/cluster-api-provider-azure/blob/7fd4f0ee12202517d5ec41a53083eba6fedfde66/api/v1beta1/azuremachine_validation.go#L140) which checks for a double slash `//` in the scope/definitionID, which would signify a subscription ID was not found. Is this a valid way to do this? I'm unable to call `GetSubscriptionID` in the validation webhook due to the absence of the client.

Reply:
How about leaving role definitionID and scope empty in `SetIdentityDefaults` if `subscriptionID` is `""`, and then simply checking those are set in the validation webhook? Checking for `//` should catch this in theory but it could lead to edge cases since that field could also be set by a user. The other thing which is arguably better is we could do is add regex validation to make sure both scope and role definitionID have valid formats. For this you could reuse `azureautorest.ParseResourceID` (although it's getting deprecated #2974, @mboersma do you know if there is an alternative?), since both are in Azure resource ID format. That would also allow us to validate user input when a user passes in a custom role.

Reply:
Hmm, looks like `azureautorest.ParseResourceID` actually returns an error for valid scope and roleDefinitionIDs, I'm trying to figure out what part of the regex doesn't match. In the meantime, I'll look for an alternative